common:
    lidar_topic:  "/points_raw"
    imu_topic:  "/imu_raw"
    gnss_topic:  "/rtk_ins"
    map_frame:  "map"
    body_frame:  "body"
    lidar_frame:  "lidar"
    timedelay_lidar2imu: 0.0 # Time offset between lidar and IMU calibrated by other algorithms, e.g. LI-Init (can be found in README).

preprocess:
    lidar_type: 2                # 1 for Livox serials LiDAR, 2 for Velodyne LiDAR, 3 for ouster LiDAR, 
    scan_line: 16
    scan_rate: 10                # only need to be set for velodyne, unit: Hz,
    timestamp_unit: 0            # the unit of time/t field in the PointCloud2 rostopic: 0-second, 1-milisecond, 2-microsecond, 3-nanosecond.
    blind: 2
    det_range: 200.0

mapping:
    gravity_align: true                 # true to align the z axis of world frame with the direction of gravity, and the gravity direction should be specified below
    gravity_init: [0, 0, -9.81]         # gravity to be aligned
    eular_init: [180, 0, 0]             # imu init rotation, for fix imu measure
    frontend_type: 0                    # frontend type

    # fastlio
    point_skip_num: 4            # for mapping, Increase accuracy 4 -> 1
    space_down_sample: true
    # surf_frame_ds_res: 0.1       # for indoor, when the ceiling features are more than the floor features after downsampling
    surf_frame_ds_res: 0.5       # In unstructured scenarios, smaller is more likely to degenerate
    max_iteration: 3
    ikdtree_resolution: 0.5      # In unstructured scenarios, smaller is more likely to degenerate
    lidar_model_search_range: 5
    lidar_meas_cov: 0.001
    cube_side_length: 1000
    runtime_log_enable: 0

    # for backend
    # keyframe_add_dist_threshold: 1
    keyframe_add_dist_threshold: 2.5
    keyframe_add_angle_threshold: 0.2
    recontruct_kdtree: true
    ikdtree_reconstruct_keyframe_num: 10
    ikdtree_reconstruct_downsamp_size: 0.1

    # for add GNSS factor
    lla: [31.4200590, 120.6394099, 14.790]
    numsv: 20
    rtk_age: 30
    gpsCovThreshold: [0.05, 0.05, 0.1, 0.05, 0.05, 0.05]
    pose_cov_threshold: 0.
    gnss_weight: [1, 1, 0.1]
    gnssValidInterval: 0.1
    useGpsElevation: true

    # for loop closure
    loop_closure_interval: 300
    loop_closure_enable_flag: true
    loop_keyframe_num_thld: 50
    loop_closure_search_radius: 40
    loop_closure_keyframe_interval: 100
    keyframe_search_num: 20
    loop_closure_fitness_score_thld: 0.2
    icp_downsamp_size: 0.1
    manually_loop_vaild_period: [0, 1]
    odom_loop_vaild_period: []
    scancontext_loop_vaild_period: []

    acc_cov: 0.1
    gyr_cov: 0.1
    b_acc_cov: 0.0001
    b_gyr_cov: 0.0001
    extrinsic_est_en:  false      # true: enable the online estimation of IMU-LiDAR extrinsic,
    # lidar -> imu = lidar->base * base->imu
    extrinsic_T: [ 1.43277653, 1.87324502, 0.04516174 ]
    extrinsic_R: [ -0.0033732, -0.99993922, -0.01048801,
                   0.99994804, -0.0034737, 0.00958088,
                   -0.00961666, -0.01045512, 0.99989908 ]
    # lidar -> gnss
    extrinsic_gnss_T: [ 0, 1.9, 1.25 ]   # cgi=car
    extrinsic_gnss_R: [ -0.0016280, -0.9999238, -0.0122332, 
                        0.9999526, -0.0017452, 0.0095778, 
                        -0.0095984, -0.0122170, 0.9998793 ]
    # extrinsic_gnss_T: [ -0.49459332, 3.43799565, 0.13195309 ]
    # extrinsic_gnss_R: [ -0.99991086, 0.00523579, -0.01228374, 
    #                     -0.00535283, -0.9999404, 0.0095129, 
    #                     -0.01223317, 0.00957778, 0.9998793 ]

scan_context:
    lidar_height: 2.183
    sc_dist_thres: 0.13

publish:
    path_en:  true
    scan_publish_en:  true       # false: close all the point cloud output
    dense_publish_en: false       # false: low down the points number in a global-frame point clouds scan.

official:
    ground_constraint_enable: false
    ground_constraint_angle: 5
    save_globalmap_en: false
    save_resolution: 0.2
    save_keyframe_en: true
    save_keyframe_descriptor_en: false
    # map_path: "/home/will/data/test_mapping/mapping1"
